### ISSUE TITLE
Service API and cancellations

### ISSUE THREAD

**Issue by pipermerriam** [ethereum/trinity#1284]
Extracted from this discussion: https://github.com/ethereum/trinity/pull/1217#discussion_r346649089

There is a foot gun that I've been pondering over for a while in the new service API.

`ManagerAPI.wait_cancelled()` is not guaranteed to ever return.

This doesn't have to be the behavior but I would like to have a discussion about it because I'm not sure which is better.

- if services **always** end in the cancelled state:
    - `wait_cancelled()` and `wait_stopped()` are mostly equivalent and only in very rare cases would using one over the other make any difference....
    - You can't easily differentiate between a service that *finished* naturally and one that was cancelled...
- if cancellation is **only** triggered by an explicit call to `Manager.cancel()` or an unhandled task error:
    - `wait_cancelled()` is generally safe to use *internally* 
    - `wait_cancelled()` called from an *external* perspective will likely not behave as the user expected because it won't ever return in the case that the service exits naturally.
    - You can easily differentiate between a service that was cancelled and a service that exited naturally.

I suspect there is a nice 3rd option that gets us the best of both worlds.  Lately I've been thinking about making a clearer differentation between the *external* manager API and the *internal* manager API (I think the *internal* manager api is a superset of the *external* API)

**Comment by ralexstokes** [comment]
I would think we want to keep the differentiation, at the least.

The fact that cancellation is somewhat exceptional, it feels more natural to me if it stays semantically distinct from "stopping".

**Comment by carver** [comment]
A nice to have might be a `cleaned_up` state that you can wait on. It would always be reached whether the service was cancelled or stopped itself.

**Comment by pipermerriam** [comment]
This is done in `async-service`.
